substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java

```diff
@@ -140,4 +216,135 @@ boolean slowlyFindPointer(Pointer p) {
     boolean walkHeapChunks(MemoryWalker.Visitor visitor) {
         return getSpace().walkHeapChunks(visitor);
     }
+
+    protected void prepareForPromotion() {
+        /* Prepare the Space walkers. */
+        for (int i = 0; i < HeapPolicyOptions.MaxSurvivors.getValue(); i++) {
+            assert getSurvivorToSpaceAt(i).isEmpty() : "SurvivorToSpace should be empty.";
+            getSurvivorGreyObjectsWalker(i).setScanStart(getSurvivorToSpaceAt(i));
+        }
+    }
+
+    protected boolean scanGreyObjects() {
+        final Log trace = Log.noopLog().string("young scanGreyObjects");
+        final GCImpl gc = HeapImpl.getHeapImpl().getGCImpl();
+        boolean needScan = false;
+
+        int maxSurvivors = HeapPolicy.getMaxSurvivors();
+
+        for (int i = 0; i < maxSurvivors; i++) {
+            if (getSurvivorGreyObjectsWalker(i).haveGreyObjects()) {
+                needScan = true;
+                break;
+            }
+        }
+
+        if (!needScan) {
+            return false;
+        }
+
+        for (int i = 0; i < maxSurvivors; i++) {
+            trace.string("[Scanning survivor-").signed(i).string("]").newline();
+            getSurvivorGreyObjectsWalker(i).walkGreyObjects(gc.getGreyToBlackObjectVisitor());
+        }
+
+        trace.flush();
+        return true;
+    }
+
+    void noteScavengeBegin() {
+        this.totalNofSurvivorChunks = 0;
+    }
+
+    void noteNewSurvivorChunk() {
+        this.totalNofSurvivorChunks += 1;
+    }
+
+    public boolean shouldPromoteToYoung(Space originalSpace) {
+        Log trace = Log.noopLog().string("[YoungGeneration::shouldPromoteYoung");
+        trace.string(", nofSurvivors=").unsigned(totalNofSurvivorChunks)
+                .string(", maxSurvivors=").unsigned(HeapPolicy.getMaxNofSurvivorChunks());
+        if (originalSpace.isYoungSpace()
+                && this.totalNofSurvivorChunks <= HeapPolicy.getMaxNofSurvivorChunks().rawValue()) {
```

Comment:
This seems to be a quite ineffective policy to decide whether to keep an object in the young generation. When the threshold is crossed while a collection is running, suddenly all remaining objects (even the ones from eden) are ending up in the old generation.

I think we should instead collect some statistics about survivor space sizes (by age) and promotion rates, and at the beginning of a collection estimate how many survivor spaces should be used. That works because the chunked heap model does not impose strict size limits on the survivor spaces. If the estimate was too optimistic, the young generation is just a bit larger than expected after a collection, but no other harm is done. The next young generation collection will then use a smaller limit for survivor spaces, i.e., promote earlier to the old generation.

And if the promotion statistics say that the promotion rate from space `n` to `n + 1` is very high (say > 80% or so), then it is also not beneficial to use space `n + 1` anymore and instead objects should be promoted to the old generation.

All of that can and should be abstracted in a separate policy class so that it is easy to experiment with different promotion strategies.